### PR TITLE
Updates some error messages in module.py and scope.oy

### LINKED ISSUE #893: Improve Error Message: Calling a method without initializing
Example code:

```
from flax import linen as nn
import jax.numpy as jnp
test_dense = nn.Dense(10)
test_dense(jnp.ones((5,5)))
```

Gives error:

```
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-4-d80093b37349> in <module>
----> 1 test_dense(jnp.ones((5,5)))

/opt/conda/envs/caliban/lib/python3.7/site-packages/flax/linen/module.py in wrapped_module_method(self, *args, **kwargs)
    146 
    147     if self.scope is None:
--> 148       raise ValueError("Can't call methods on orphaned modules")
    149 
    150     if is_compact_method:

ValueError: Can't call methods on orphaned modules
```

The problem is that you should run `init` or `apply`, but the error message does not specify this.

This code does work:

```
# create the initialized variables w. a random key for inits:
variables = test_dense.init(random.PRNGKey(0), jnp.ones((5,5)))
# apply the NN to the variables + input to get output
y = test_dense.apply(variables, jnp.ones((5,5)))
```

### PR DESCRIPTION
**Pull request by marcvanzee**
Closes #893

### CODE REVIEW

File: flax/errors.py [hunk 1]
```diff
@@ -117,4 +116,44 @@ def __call__(self, x):
...
+  this, add this to your imports::
+    
+    from jax.config import config
+    config.enable_omnistaging()
+  """
+  def __init(self):
+    super().__init__(f'Flax Linen requires Omnistaging to be enabled')
+
+
+class UnboundCompactModuleError(FlaxError):
```
Comment (avital) [inline flax/errors.py:134]: This new error alone is so valuable

File: flax/errors.py [hunk 2]
```diff
@@ -117,4 +116,44 @@ def __call__(self, x):
...
+
+    test_dense = nn.Dense(10)
+    test_dense(jnp.ones((5,5)))
+
+  Instead, you should pass the variables (parameters and other state) via 
+  :meth:`Module.apply() <flax.linen.Module.apply>` (or use 
+  :meth:`Module.init() <flax.linen.Module.init>` to get initial variables)::
+
+    # Create the initialized variables with a random key for inits:
+    params = test_dense.init(random.PRNGKey(0), jnp.ones((5,5)))['params']
```
Comment (avital) [inline flax/errors.py:151]: I think this code doesn't work. It should be `variables = ` and don't do `['params']` at the end

File: flax/linen/module.py [hunk 3]
```diff
@@ -231,7 +228,7 @@ def wrapped_module_method(self, *args, **kwargs):
 
     if is_compact_method:
       if self.scope is None:
-        raise ValueError("Can't call compact methods on unbound modules")
+        raise UnboundCompactModuleError(self.name)
```
Comment (avital) [inline flax/linen/module.py:231]: I'm curious whether we also get this error when calling non-compact methods

File: flax/errors.py (hunk 2 above)
Comment (avital) [reply flax/errors.py:151]: Ah, I get a different error then -- `setup` doesn't run so the submodule I defined isn't there

https://colab.research.google.com/drive/1SsRzkBoGqf1DNYRVseh5QxjA_SwCb7gb

Comment (marcvanzee) [reply flax/errors.py]: I don't seem to get a different error when I do as you suggested:

```
from flax import linen as nn
import jax.numpy as jnp

test_dense = nn.Dense(10)

 # Create the initialized variables with a random key for inits:
import jax.random
vars = test_dense.init(random.PRNGKey(0), jnp.ones((5,5)))

# Apply the NN to the variables + input to get output.
y = test_dense.apply(vars, jnp.ones((5,5)))
```

File: flax/linen/module.py (hunk 3 above)
Comment (marcvanzee) [reply flax/linen/module.py]: It doesn't seem to be possible since it states `if is_compact_pmethod` before, right?

Comment (avital) [reply flax/linen/module.py]: Right, I mean that it would be good to similarly detect that -- this was independently filed as #1014

**Comment by marcvanzee** [comment]
Closing this since I've created another PR #1070 which uses my local branch.

File: flax/linen/module.py (hunk 3 above)
Comment (marcvanzee) [reply flax/linen/module.py]: Good idea, I've moved your suggestion to the description of the new PR #1070 and I'll implement it there.

**Pull request closed without merging**